### PR TITLE
[RFR] Rewrote the whole library... :x

### PR DESCRIPTION
**Pull request by HugoGiraudel**
At first I wanted to make the expression parser a bit more flexible. Then I wanted to make the media inclusion a bit less hacky. One thing after another, I believe I have rewritten most of the library... If you don't like it @eduardoboucas, feel free to decline the PR. 

## Changes

* More elegant and flexible expression parser (found in `./src/helpers/_parser.scss`)
* Recursive mixin approach relying on Sass media merging rather than complex Sass logic (found in `./src/helpers/_outputter.scss`)
* Use of `@error` while still supporting Sass 3.3 with a custom logger switching between `@error` and `@warn` (found in `./src/helpers/_logger.scss`)
* Changed some tabs to spaces

## Breaking changes

As far as I can tell, there is none. The API still works like before, it's only about the technical implementation.

### CODE REVIEW

File: dist/_include-media.scss [hunk 1]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
+
+  @return $message;
+}
+
+/// 
+/// Wrapper mixin for the log function so it can be used with a more friendly
+/// API than `@if log('..') {}` or `$_: log('..')`. Basically, use the function
+/// within functions because it is not possible to include a mixin in a function
+/// and use the mixin everywhere else because it's much more elegant.
+///
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Missing `@access`.
Missing `@param`.

File: dist/_include-media.scss [hunk 2]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
 ///
-/// @example scss - With a single set breakpoint
-///  @include media(">phone") { }
+/// Internal recursive mixin for include-media
+/// 
+/// @access private
 ///
-/// @example scss - With two set breakpoints
-///  @include media(">phone", "<=tablet") { } 
+/// @param {List} $queries - List of parsed queries
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Missing `@see`.

File: dist/_include-media.scss [hunk 3]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
+@function get-expression-operator($expression) {
+  @each $operator in ('>=', '>', '<=', '<', '≥', '≤') {
+    @if str-index($expression, $operator) {
+      @return $operator;
+    }
   }
 
-  $branches: get-query-branches($list);
-  $query: '';
+  $_: log('No operator found in `#{$expression}`.');
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Comment this as it's non-intuitive.

File: dist/_include-media.scss [hunk 4]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
+/// @return {String} - `width` or `height` (or potentially anything else)
+///
+@function get-expression-dimension($expression, $operator) {
+  $operator-index: str-index($expression, $operator);
+  $dimension: str-slice($expression, 0, $operator-index - 1);
 
-  @media #{$query} {
-    @content;
-  }
+  @return if(str-length($dimension) == 0, 'width', $dimension);
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Not elegant, to be document at worst, changed at best.

File: dist/_include-media.scss [hunk 5]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
+  $value: str-slice($expression, $operator-index + str-length($operator));
 
-  // Getting initial snapshot and looking for groups
-  @each $expression in $expressions {
-    @if (str-length($result) != 0) {
-      $result: $result + ' and ';
-    }
+  @if map-has-key($breakpoints, $value) {
+    $value: map-get($breakpoints, $value);
+  } @else if map-has-key($media-expressions, $value) {
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Can't happen. Remove this.

File: dist/_include-media.scss [hunk 6]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
-    } @else if (type-of($expression) == 'list') {
-      $result: $result + nth($expression, 1);
-      $has-groups: true;
-    } @else {
-      @warn '#{$expression} is not a valid expression.';
-    }
+  $interval: map-get($unit-intervals, unit($value));
+
+  @if not $interval {
+    $_: log('Unknown unit `#{unit($value)}`.');
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Comment this as it's non-intuitive.

File: dist/_include-media.scss [hunk 7]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
-        @each $member in $expression {
-          @if ($member != $first) {
-            @each $partial in $result {
-              $result: join($result, str-replace-first($first, $member, $partial));
-            }
-          }
-        }
-      }
-    }
+  @if $operator == '>' {
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Make this a bit more elegant maybe?

File: dist/_include-media.scss [hunk 8]
```diff
@@ -65,230 +66,251 @@ $media-expressions: (
...
 ///
-/// Generates a media query based on a list of conditions
+/// Log a message either with `@error` if supported 
+/// else with `@warn`, using `function-exists('at-error')`
+/// to detect support.
+/// 
+/// @param {String} $message - Message to log
+///
+@function log($message) {
+  // `== true` is needed for Sass 3.2 which has no support
```
Comment (HugoGiraudel) [inline dist/_include-media.scss]: Sass 3.2 is not supported, so remove all this stuff.

**Comment by eduardoboucas** [comment]
All looks good to me!

**Comment by eduardoboucas** [comment]
I found another issue.

SCSS:
```scss
@include media('landscape', '>phone', '<tablet') {
  color: tomato;
}
```

Error:
```
Invalid CSS after "": expected media query list, was ""(orientation: ..."
```

Using Ruby Sass 3.4.13 and 3.4.5.

**Comment by eduardoboucas** [comment]
I think has now been fixed in #46 as well. First of all, we no longer need to declare media expressions with logical disjunctions as maps, they can be simple strings. Also, by adding `unquote()` it works with media expressions for me, gives me an error otherwise.

**Pull request merged**